bokeh/client/session.py

```diff
@@ -306,11 +323,23 @@ def _document_changed(self, event):
             return
 
         if isinstance(event, SessionCallbackAdded):
-            self._add_periodic_callback(event.callback)
+            if isinstance(event.callback, PeriodicCallback):
+                self._add_periodic_callback(event.callback)
+            elif isinstance(event.callback, TimeoutCallback):
+                self._add_timeout_callback(event.callback)
+            else:
+                raise ValueError("Expected callback of type PeriodicCallback or TimeoutCallback, got: %s" % event.callback)
+
             return
 
-        if isinstance(event, SessionCallbackRemoved):
-            self._remove_periodic_callback(event.callback)
+        elif isinstance(event, SessionCallbackRemoved):
+            if isinstance(event.callback, PeriodicCallback):
+                self._remove_periodic_callback(event.callback)
+            elif isinstance(event.callback, TimeoutCallback):
+                self._remove_timeout_callback(event.callback)
+            else:
+                raise ValueError("Expected callback of type PeriodicCallback or TimeoutCallback, got: %s" % event.callback)
+
```

Comment:
It bothers me to have knowledge of Session inside `document.py` or inside the events. This is a multiple-dispatch problem I guess (http://python-3-patterns-idioms-test.readthedocs.org/en/latest/MultipleDispatching.html ). One solution could be this (untested but hopefully you see what I mean):

```
class DocumentChangedEvent(object):
    def dispatch(self, receiver):
        if hasattr(receiver, '_document_changed'):
          receiver._document_changed(self)

class SessionCallbackAdded(DocumentChangedEvent):
    def dispatch(self, receiver):
        super(SessionCallbackAdded, self).dispatch(receiver)
        if hasattr(receiver, '_session_callback_added'):
          receiver._session_callback_added(self)
```

Receiver would have to attach like this I guess:
```
 document.on_change(lambda event: event.dispatch(self))
```

That could maybe be a convenience method on Document, like:
```
class Document:
  def on_change_dispatch_to(self, receiver):
     self.on_change(lambda event: event.dispatch(receiver))
```

?